File: integrations/shared/native/TriggerTests.cpp
#include "StudioTriggerManager.h"
#include <iostream>

using namespace MetroidvaniaStudio;

int main(int argc, char **argv)
{
    try
    {
        Require(argc == 2, "Pass the trigger fixture JSON.");
        const auto document = JsonReader(ReadText(std::filesystem::u8path(argv[1]))).Read();
        Room room;
        room.metadata = document.Get("rooms").Items().at(0);
        room.id = room.metadata.Get("id").Text();
        room.name = room.metadata.Get("name").Text();
        room.x = room.metadata.Get("x").Integer();
        room.y = room.metadata.Get("y").Integer();
        TriggerManager manager;
        manager.RegisterRoom(room);
        TriggerInfo info;
        int delivered = 0;
        manager.onTriggered = [&](const TriggerInfo &value)
        {
            ++delivered;
            if (value.objectId == "once")
            {
                TriggerInfo recursive;
                Require(!manager.TryRequest("once", recursive), "Reentrant once request.");
            }
        };
        Require(manager.TryGet("once", info) && info.once, "Query should not consume.");
        Require(manager.TryRequest("once", info) && info.roomX == -7 && info.roomY == 3 && info.event == TriggerEvent::Trigger200 && info.description == "Open gate", "Delivery payload.");
        Require(!manager.TryRequest("once", info), "Once delivers once.");
        Require(manager.TryRequest("repeat", info) && manager.TryRequest("repeat", info), "Repeat delivers repeatedly.");
        Require(manager.TryRequest("portal", info) && info.once && !manager.TryRequest("portal", info), "Portal always one-shot.");
        Require(!manager.TryRequest("legacy", info) && !manager.TryRequest("absent", info), "Unassigned requests reject.");
        room.x = 10;
        manager.RegisterRoom(room);
        Require(!manager.TryRequest("once", info) && manager.TryGet("once", info) && info.roomX == 10, "Reload preserves consumed state.");
        manager.UnregisterRoom(room.id);
        manager.RegisterRoom(room);
        Require(!manager.TryRequest("once", info), "Reenter retains once state.");
        auto other = room;
        other.id = "other";
        bool rejected = false;
        try { manager.RegisterRoom(other); } catch (const std::exception &) { rejected = true; }
        Require(rejected && manager.TryGet("once", info) && info.roomId == room.id, "Duplicate registration is atomic.");
        Require(manager.ResetOnce("once") && manager.TryRequest("once", info), "Reset reactivates.");
        manager.ResetAllOnce();
        Require(manager.TryRequest("portal", info), "Reset all reactivates.");
        for (int i = 1; i <= 200; ++i)
        {
            const auto number = std::to_string(i);
            Require(static_cast<int>(ParseTriggerEvent("Trigger" + std::string(3 - number.size(), '0') + number)) == i, "Stable enum numbers.");
        }
        for (const auto &value : {"old", "Trigger201", "Trigger1", "-1", "+1", "1.0", "999999999999999999"})
            Require(ParseTriggerEvent(value) == TriggerEvent::None, "Invalid event.");
        manager.Clear();
        Require(!manager.TryGet("once", info), "Clear removes state.");
        Require(delivered == 6, "Expected six successful callbacks.");
        std::cout << "Trigger requests passed: payload, 200 events, once, portal, reset, reentry and atomic duplicate rejection.\n";
        return 0;
    }
    catch (const std::exception &error)
    {
        std::cerr << error.what() << '\n';
        return 1;
    }
}

File: metroidvania-studio/core-tests/TriggerTests.cs
using MetroidvaniaStudio;

internal static class TriggerTests
{
    private static void Check(bool value, string message) { if (!value) throw new Exception(message); }
    public static void Requests()
    {
        foreach (int i in Enumerable.Range(0, 201))
        {
            string name = i == 0 ? "None" : "Trigger" + i.ToString("D3");
            Check(StudioTriggerEvents.TryParse(name, out var parsed) && (int)parsed == i, "Stable event number: " + name);
        }
        foreach (string invalid in new[] { "old-event", "Trigger201", "Trigger1", "-1", "+1", "1.0", "999999999999999999999" })
            Check(!StudioTriggerEvents.TryParse(invalid, out _), "Reject unsupported event: " + invalid);
        MapObject Object(string id, bool once, string definition = "Area", string triggerEvent = "Trigger200") => new()
        {
            id = id, definition = definition, layer = definition == "Area" ? MapLayer.Triggers : MapLayer.Entities,
            properties = new() { new() { key = "event", value = triggerEvent }, new() { key = "once", value = once.ToString() }, new() { key = "desc", value = "Open gate / 문 열기" } }
        };
        var room = new MapRoom { id = "room-a", name = "Entry", x = -7, y = 3,
            objects = new() { Object("repeat", false), Object("once", true), Object("portal", false, "Portal"), Object("wall", false, "InvisibleWall"), Object("legacy", false, "Area", "old-event") } };
        var manager = new StudioTriggerManager(); manager.RegisterRoom(room);
        var deliveries = new List<StudioTriggerInfo>();
        manager.TriggerRequested += info =>
        {
            deliveries.Add(info);
            if (info.ObjectId == "once") Check(!manager.TryRequest(info.ObjectId, out _), "One-shot callback cannot reenter.");
        };
        Check(manager.TryGet("once", out var before) && before.Once, "Query does not consume.");
        Check(manager.TryRequest("once", out var info) && !manager.TryRequest("once", out _), "Once delivers once.");
        Check(info.RoomX == -7 && info.RoomY == 3 && info.Event == StudioTriggerEvent.Trigger200 && info.Description == "Open gate / 문 열기", "Authored coordinates, enum, ID and description survive.");
        Check(manager.TryRequest("repeat", out _) && manager.TryRequest("repeat", out _), "Repeat events repeat.");
        Check(!manager.TryGet("portal", out _) && !manager.TryRequest("portal", out _) && !manager.TryRequest("wall", out _), "Portals and invisible walls never dispatch numbered events, including legacy properties.");
        Check(!manager.TryRequest("legacy", out _) && !manager.TryRequest("missing", out _), "No accidental dispatch for unassigned or missing events.");
        room.x = 100; room.objects[1].properties[2].value = "Changed";
        Check(manager.TryGet("once", out info) && info.RoomX == -7, "Registration stores a snapshot.");
        manager.RegisterRoom(room);
        Check(!manager.TryRequest("once", out _) && manager.TryGet("once", out info) && info.RoomX == 100, "Reload updates metadata but retains once state.");
        manager.UnregisterRoom(room.id); manager.RegisterRoom(room);
        Check(!manager.TryRequest("once", out _), "Reentering a room retains once state.");
        var duplicate = new MapRoom { id = "room-b", objects = new() { Object("repeat", false) } };
        bool rejected = false; try { manager.RegisterRoom(duplicate); } catch (ArgumentException) { rejected = true; }
        Check(rejected && manager.TryGet("repeat", out info) && info.RoomId == room.id, "Duplicate IDs reject atomically.");
        duplicate.objects[0].id = ""; rejected = false;
        try { manager.RegisterRoom(duplicate); } catch (ArgumentException) { rejected = true; }
        Check(rejected, "Empty IDs reject.");
        Check(manager.ResetOnce("once") && manager.TryRequest("once", out _), "Explicit reset reactivates.");
        manager.ResetAllOnce(); Check(manager.TryRequest("once", out _) && !manager.TryRequest("portal", out _), "Reset only reactivates event triggers.");
        manager.Clear(); Check(!manager.TryGet("once", out _), "Clear removes data and session state.");
        manager.RegisterRoom(room); Check(manager.TryRequest("once", out _), "New world starts unconsumed.");
    }
}
